### PR TITLE
Fix/fix broadcast overlap with isp

### PR DESCRIPTION
**Pull request by mwiacx**
insure that parameter update by async broadcast will be done before isp prefetch parameter.

### CODE REVIEW

File: internlm/core/communication/isp.py [hunk 1]
```diff
@@ -307,7 +312,14 @@ def _all_gather_module_weight(self, module):
         self._weight_global_output[module] = weight_output
 
     def _all_gather_block_weight(self, block_index: int):
-        for module in self._index_to_isp_module[block_index]:
+        block = self._index_to_block[block_index]
+
+        # wait for prerequisite conditions
+        for callback in self._forward_prefetch_prerequisites:
```
Comment (huangting4201) [inline internlm/core/communication/isp.py]: 如果开了ckpt并且self.is_forward = False的话，是不是不需要执行callback函数

File: internlm/core/communication/utils.py [hunk 2]
```diff
@@ -123,3 +129,100 @@ def gather_split_1d_tensor(tensor: torch.Tensor) -> torch.Tensor:
...
+
+        # register_forward_pre_hook for transformer/embeding/norm/xxx block
+        for block, _ in self._block_to_rank.items():
+            # TODO: remove special handling for embedding and head layers,
+            # instead implement support for weight parallelism of embedding and head layers within the ISP.
+
+            # NOTE: Although the layernorm layer does not have explicit processing,
+            # both ISPCommunicator and ParamAsyncBcastHandler handle transformer blocks as granularity,
+            # so everything is fine.
+            if isp_communicator is None or isinstance(block, (Embedding1D, ScaleColumnParallelLinear)):
```
Comment (huangting4201) [inline internlm/core/communication/utils.py]: embedding and head
isinstance(module, (Embedding1D, ParallelGPT2Embeddings, BaseScaleColumnParallelLinear)):
换成这样是不是更保险一些

File: internlm/core/communication/isp.py (hunk 1 above)
Comment (mwiacx) [reply internlm/core/communication/isp.py]: 理论上不影响，不过还是写一写比较好一些

File: internlm/core/communication/utils.py (hunk 2 above)
Comment (mwiacx) [reply internlm/core/communication/utils.py]: 可以，我加一下

File: internlm/core/communication/isp.py (hunk 1 above)
Comment (mwiacx) [reply internlm/core/communication/isp.py]: [a5c5dfa](https://github.com/InternLM/InternEvo/pull/64/commits/a5c5dfa03560c3699715e3ea61e8f0b62a8b7be1)

File: internlm/core/communication/utils.py (hunk 2 above)
Comment (mwiacx) [reply internlm/core/communication/utils.py]: [a5c5dfa](https://github.com/InternLM/InternEvo/pull/64/commits/a5c5dfa03560c3699715e3ea61e8f0b62a8b7be1)

**Review by sunpengsdu — APPROVED** [review]
LGTM

**Pull request merged**